### PR TITLE
Support case-insensitive parquet nested column name-based look up

### PR DESCRIPTION
**Pull request by Yaliang**
The existing tests for optimized parquet reader only cover the cases using index-based parquet column lookup(a.k.a `useParquetColumnNames` is false) . And `testComplexNestedStructs` will be failed because the deeply nested columns contain uppercase characters. This patch can fix this error.
This patch has been verified in production for a few days. 

In addition, adding the `parquet_use_column_names` to override the default setting `hive.parquet.use-column-names` in a session base gives more convenience in testing and real use case.

### CODE REVIEW

**Comment by Yaliang** [comment]
@kbajda @zhenxiao @nezihyigitbasi 

Could you take a review?

**Comment by nezihyigitbasi** [comment]
There are checkstyle failures @Yaliang.

**Comment by Yaliang** [comment]
@nezihyigitbasi Thanks to pointing that out. Fixed!

**Comment by Yaliang** [comment]
@zhenxiao 

1. The patch is actually a fix to the current optimized reader. Because the Hive is case insensitive, if you enable both the optimized reader and use column names at the same time, the query will fail with NPE if the nested column names from parquet file contain upper case.

2. I'm adding the session property so that the configuration with be session based. It would be easier to switch in the tests. The default value of such will be given by setting hive.parquet.use-column-names.

**Comment by zhenxiao** [comment]
get it. thanks for the explanation @Yaliang 

if setting hive.parquet.use-column-names with new Parquet Reader triggered NPE for upper case field names, I am inclined to fix the behavior for hive.parquet.use-column-names, to make it case insensitive, other than introducing another configuration. What do you think?

**Comment by Yaliang** [comment]
@zhenxiao 

I may not explain properly. `hive.parquet.use-column-names` is configuration coming from a configuration file. The session property `parquet_use_column_names` allows user to override that configuration. Similar to `force_local_scheduling` vs `hive.force-local-scheduling`.

**Comment by zhenxiao** [comment]
oh, get it. I am good with adding the session property.

**Comment by Yaliang** [comment]
Kindly ping for reviews

@kgalieva @zhenxiao @nezihyigitbasi

**Comment by Yaliang** [comment]
Rebased

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHiveFileFormats.java [hunk 1]
```diff
@@ -785,6 +786,15 @@ private FileFormatAssertion assertThatFileFormat(HiveStorageFormat hiveStorageFo
                 .withStorageFormat(hiveStorageFormat);
     }
 
+    private static HiveClientConfig getParquetHiveClientConfig(boolean optimizedReader, boolean usePushDown, boolean useParquetColumnNames)
```
Comment (nezihyigitbasi) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestHiveFileFormats.java]: `getParquetHiveClientConfig` -> `createParquetHiveClientConfig`

Comment (nezihyigitbasi) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestHiveFileFormats.java]: To be consistent with the methods names in `HiveClientConfig`:
- `optimizedReader` -> `enableOptimizedReader`
- `usePushDown` -> `enablePredicatePushDown`

Comment (nezihyigitbasi) [reply presto-hive/src/test/java/com/facebook/presto/hive/TestHiveFileFormats.java]: You can chain calls:

```
HiveClientConfig config = new HiveClientConfig();
config.setParquetOptimizedReaderEnabled(optimizedReader)
        .setParquetPredicatePushdownEnabled(usePushDown)
        .setUseParquetColumnNames(useParquetColumnNames);
return config;
```

File: presto-hive/src/test/java/com/facebook/presto/hive/parquet/ParquetTester.java [hunk 2]
```diff
@@ -383,12 +394,13 @@ private static List toRowValue(Block rowBlock, List<Type> fieldTypes)
         return Collections.unmodifiableList(values);
     }
 
-    private static HiveClientConfig getHiveClientConfig()
+    private static HiveClientConfig getHiveClientConfig(boolean useParquetColumnNames)
```
Comment (nezihyigitbasi) [inline presto-hive/src/test/java/com/facebook/presto/hive/parquet/ParquetTester.java]: To be consistent with the other similar method, you can rename this as `createHiveClientConfig` and chain the method calls in the body. Please do this refactor in a separate commit.

File: presto-hive/src/main/java/com/facebook/presto/hive/parquet/ParquetTypeUtils.java [hunk 3]
```diff
@@ -246,6 +246,25 @@ public static ParquetEncoding getParquetEncoding(Encoding encoding)
         return null;
     }
 
+    public static ColumnIO getColumnIOByName(GroupColumnIO groupColumnIO, String name)
```
Comment (nezihyigitbasi) [inline presto-hive/src/main/java/com/facebook/presto/hive/parquet/ParquetTypeUtils.java]: Let's add a javadoc comment here (and remove the comments below in the method body):

```
/**
 * Parquet column names are case-sensitive unlike Hive, which converts all column names to lowercase.
 * Therefore, when we look up columns we first check for exact match, and if that fails we look for a case-insensitive match.
 */
```

File: presto-hive/src/main/java/com/facebook/presto/hive/parquet/ParquetPageSource.java [hunk 4]
```diff
@@ -105,7 +106,7 @@ public ParquetPageSource(
             }
             else {
                 String columnName = useParquetColumnNames ? name : fileSchema.getFields().get(column.getHiveColumnIndex()).getName();
-                fieldsBuilder.add(constructField(type, messageColumnIO.getChild(columnName)));
+                fieldsBuilder.add(constructField(type, getColumnIOByName(messageColumnIO, columnName)));
```
Comment (nezihyigitbasi) [inline presto-hive/src/main/java/com/facebook/presto/hive/parquet/ParquetPageSource.java]: `getColumnIOByName` can return `null`, and that will cause a NPE here in `add` call. If `getColumnIOByName` returns `null` it means a column with given name couldn't be found, and we get the column name from the Parquet file schema, so I feel like that's a serious error condition and instead of returning a null we should throw.

File: presto-hive/src/main/java/com/facebook/presto/hive/parquet/ParquetTypeUtils.java (hunk 3 above)
Comment (nezihyigitbasi) [inline presto-hive/src/main/java/com/facebook/presto/hive/parquet/ParquetTypeUtils.java]: `getColumnIOByName` -> `lookupColumnByName`

**Comment by nezihyigitbasi** [comment]
@Yaliang For the next review round, please squash the commits to have only two commits: add a new session property and support for case-insensitive column lookup.

File: presto-hive/src/main/java/com/facebook/presto/hive/parquet/ParquetPageSource.java (hunk 4 above)
Comment (Yaliang) [reply presto-hive/src/main/java/com/facebook/presto/hive/parquet/ParquetPageSource.java]: I don't know which one is better. I feel that, in the most case, if the field missed in Parquet, it seems OK to return `null`. For example, the `getParquetType` may also return `null`. How do you think?

**Comment by Yaliang** [comment]
@nezihyigitbasi Thanks for your review. I have updated the commit. I don't know whether we should throw an exception in case the column missed. I would prefer just keep it as is because it's consistent with other methods, for example, `getParquetType`.

**Comment by nezihyigitbasi** [comment]
@Yaliang there are checkstyle failures.

**Comment by nezihyigitbasi** [comment]
@Yaliang thanks, merged!
